pkg/provider/consulcatalog/consul_catalog.go

```diff
@@ -294,6 +346,136 @@ func contains(values []string, val string) bool {
 	return false
 }
 
+func rootsWatchHandler(ctx context.Context, dest chan<- []string) func(watch.BlockingParamVal, interface{}) {
+	return func(_ watch.BlockingParamVal, raw interface{}) {
+		if raw == nil {
+			return
+		}
+
+		v, ok := raw.(*api.CARootList)
+		if !ok || v == nil {
+			log.FromContext(ctx).Errorf("invalid result for root certificate watcher")
+			return
+		}
+
+		roots := make([]string, 0, len(v.Roots))
+		for _, root := range v.Roots {
+			roots = append(roots, root.RootCertPEM)
+		}
+
+		dest <- roots
+	}
+}
+
+type keyPair struct {
+	cert string
+	key  string
+}
+
+func leafWatcherHandler(ctx context.Context, dest chan<- keyPair) func(watch.BlockingParamVal, interface{}) {
+	return func(_ watch.BlockingParamVal, raw interface{}) {
+		if raw == nil {
+			return
+		}
+
+		v, ok := raw.(*api.LeafCert)
+		if !ok || v == nil {
+			log.FromContext(ctx).Errorf("invalid result for leaf certificate watcher")
+			return
+		}
+
+		dest <- keyPair{
+			cert: v.CertPEM,
+			key:  v.PrivateKeyPEM,
+		}
+	}
+}
+
+func (p *Provider) watchConnectTLS(ctx context.Context) {
```

Comment:
It is problematic that watchConnectTLS could in theory fail at different points, without the caller being aware about it at all. Then Provide would keep on running as if everything's normal even though it will never see any certificate renewal and will never trigger the corresponding configuration refresh.
So there's two ways we could go about this:

1) Make watchConnectTLS more resilient, with more loops, handlers getting restarted when they fail, etc. But that does not seem appropriate to me as it seems most of the potential errors here are critical, i.e. retrying things would just end up failing in the same way.

2) Make watchConnectTLS report to its caller when anything fails. So one quick fix we could do is add an err field to connectCert, and whenever there's a critical error in there (instead of logging), we send a connectCert, with the corresponding error set, in p.certChan, and we return. And it's up to the caller (Provide) to check that error field, and  log and fail gracefully when needed. WDYT?

And in Provide, if an error from watchConnectTLS is detected, then the operation should return a backoff.Permanent error so that we can break out of the backoff.RetryNotify.

Reply:
Good points, I'll have to dig more closely into the source code. In theory `RunWithClientAndHclog` https://github.com/hashicorp/consul/blob/f0872a4ac1a0d89492a2829666e2d408cda81f96/api/watch/plan.go#L59 can return an error, but I only see it returning nil -- seems as if it would just backoff. I'll have to check the failures in plan creation.

Reply:
Please see the last commit. I tried to separate unrecoverable errors from the rest and reworked it into two functions. This is probably as far as my go knowledge can get us. :/